### PR TITLE
utp_shutdown

### PR DESCRIPTION
**Pull request by ghazel**
`utp_shutdown()` for half-close uTP semantics like TCP.

### CODE REVIEW

File: utp_internal.cpp [hunk 1]
```diff
@@ -3363,22 +3381,62 @@ void utp_close(UTPSocket *conn)
...
+		} else if (conn->fin_sent_acked) {
+			conn->state = CS_DESTROY;
+		}
 		break;
 
 	case CS_SYN_SENT:
 		conn->rto_timeout = utp_call_get_milliseconds(conn->ctx, conn) + min<uint>(conn->rto * 2, 60);
 		// fall through
-	case CS_GOT_FIN:
-		conn->state = CS_DESTROY_DELAY;
```
Comment (ssiloti) [inline utp_internal.cpp]: This was the only place `CS_DESTROY_DELAY` was being entered. Is this state still needed?

Comment (ghazel) [reply utp_internal.cpp]: Nope, not needed. Want me to remove it?

Comment (ssiloti) [reply utp_internal.cpp]: Yes please.

**Comment by ghazel** [comment]
Removed `CS_DESTROY_DELAY`.

**Comment by anacrolix** [comment]
Do you have an example where shutdown is needed, or preferred? I have a [Go wrapper](https://github.com/anacrolix/go-libutp) and I wonder if there's any areas that will be made more reliable by using shutdown semantics.

**Comment by ghazel** [comment]
When trying to make an accurate socket proxy, shutdown is sometimes required. HTTP clients in particular shutdown to indicate the last request is sent, but they still expect a reply.

**Pull request merged**